library/dnet.c

```diff
@@ -671,6 +672,28 @@ static int dnet_server_send_sync(struct dnet_server_send_ctl *ctl)
 	return dnet_server_send_put(ctl);
 }
 
+static int dnet_filter_invalid_groups(int *groups, const struct dnet_server_send_ctl *send, struct dnet_node *n, struct dnet_id *id)
```

Comment:
This requires more comments about the whole logic behind this function.
Why is it needed, what does it do, how does it modify group array.

One should understand this function without looking into the code, but from description.
Looking into the code is only needed when you want to fix a bug.

It also leaks state.